daemon/exec.go

```diff
@@ -292,16 +293,15 @@ func (daemon *Daemon) ContainerExecStart(ctx context.Context, name string, optio
 
 	select {
 	case <-ctx.Done():
+		sigCtx, cancel := context.WithTimeout(context.Background(), termProcessTimeout)
+		defer cancel()
 		logrus.Debugf("Sending TERM signal to process %v in container %v", name, c.ID)
-		daemon.containerd.SignalProcess(ctx, c.ID, name, signal.SignalMap["TERM"])
-
-		timeout := time.NewTimer(termProcessTimeout)
-		defer timeout.Stop()
+		daemon.containerd.SignalProcess(sigCtx, c.ID, name, signal.SignalMap["TERM"])
 
 		select {
-		case <-timeout.C:
+		case <-sigCtx.Done():
 			logrus.Infof("Container %v, process %v failed to exit within %v of signal TERM - using the force", c.ID, name, termProcessTimeout)
-			daemon.containerd.SignalProcess(ctx, c.ID, name, signal.SignalMap["KILL"])
+			daemon.containerd.SignalProcess(context.Background(), c.ID, name, signal.SignalMap["KILL"])
```

Comment:
`SignalProcess` sends an RPC, therefore could potentially hang for a long time if things go wrong. Some timeout might be appropriate?

I wish execs were terminated by calling a method rather than by canceling a context. A caller would then be able to pass a dedicated context just for the termination operation instead of having to hardcode some policy into the `ContainerExecStart` implementation.

Reply:
Makes sense to me to use a timeout here, but which value to use? `termProcessTimeout` applies to target process to terminate, not internal communication delays.

About global design, I agree use of a cancelable context here to stop exec is not very pleasant, considering the significant amount of additional API calls involved.

Reply:
I'd pick a timeout value which is large enough to never be hit when containerd is responsive, accounting for potentially massive network latencies, but which is not actually forever so that requests to an unresponsive containerd (or a [slowloris](https://en.wikipedia.org/wiki/Slowloris_(computer_security)) honeypot masquerading as a containerd) don't leak resources indefinitely. How does 30 seconds each sound?